### DISCUSSION TITLE
WSL2 fails to start with networkingMode=bridged

### DISCUSSION THREAD

**Discussion by Kilill** [microsoft/WSL#9227] [General] [answered]
### Version

Microsoft Windows [Version 10.0.19045.2311]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

Linux version 5.15.74.2-microsoft-standard-WSL2 (oe-user@oe-host) (x86_64-msft-linux-gcc (GCC) 9.3.0, GNU ld (GNU Binutils) 2.34.0.20200220) #1 SMP Wed Nov 2 19:50:29 UTC 2022

### Distro Version

 Ubuntu 20.04.3 LTS

### Repro Steps

create a virtual switch according to : [(https://github.com/luxzg/WSL2-fixes/blob/master/networkingMode%3Dbridged.md#install-wsl-2-preview)]

fresh install of wsl:
`wsl --install`
terminated and shutdown wsl

add the lines:
```
[wsl2]
networkingMode=bridged
vmSwitch=WSL_external
```
to `.wslconfig`

try runnig wsl

### Expected Behavior

wsl should start

### Actual Behavior

wsl command terminates with:

```
The virtual machine or container JSON document is invalid.
Error code: Wsl/Service/CreateInstance/CreateVm/ConfigureNetworking/0x8037010d
```

removing the `networkingMode=bridged` line from the config allows wsl to start

**Comment by elsaco** [comment]
@Kilill WSL network bridging, as per mentioned guide, works in Windows 11 only. Looks like you're using Windows 10.

**Reply by Anutrix** [reply]
I have the same issue on Windows 11.

**Comment by Kilill** [comment]
Yeah noticed that, but afaik that guide was also done for the prerelease of WSL 2 before it transitioned into V1.0 and supposedly supported on W10...

**Comment by elsaco** [comment]
I have it working on Windows 11 but on Windows 10 it keeps throwing errors:

```
The virtual machine or container JSON document is invalid.
Error code: Wsl/Service/CreateInstance/CreateVm/ConfigureNetworking/0x8037010d
The virtual machine or container JSON document is invalid.
Error code: Wsl/Service/CreateInstance/CreateVm/ConfigureNetworking/0x8037010d
The virtual machine or container JSON document is invalid.
Error code: Wsl/Service/CreateInstance/CreateVm/ConfigureNetworking/0x8037010d
Press any key to continue...
```

WSL version used for testing:

```
PS C:\Users\elsaco> wsl --version
WSL version: 1.0.0.0
Kernel version: 5.15.74.2
WSLg version: 1.0.47
MSRDC version: 1.2.3575
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.2311
```

**Reply by osde8info** [reply]
getting same error

```
The virtual machine or container JSON document is invalid.
Error code: Wsl/Service/CreateInstance/CreateVm/ConfigureNetworking/0x8037010d

[process exited with code 4294967295 (0xffffffff)]
```

are there any plans to enable hyperv vswitches on windows 10 ?

```
wsl --version
WSL version: 1.0.3.0
Kernel version: 5.15.79.1
WSLg version: 1.0.47
MSRDC version: 1.2.3575
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.2364
```

**Comment by Balder1840** [comment]
got same issue. Any workaround plz?

**Comment by jieny** [comment]
got same issue. Any workaround plz?

**Reply by osde8info** [reply]
the only workaround ive found is to use different port numbers in each vm

**Reply by jieny** [reply]
> the only workaround ive found is to use different port numbers in each vm

mee too.
I Forward host 80 port to wsl 8080

**Comment by gtianw** [comment]
got same issue. Any workaround plz?

**Comment by mrzhang76** [comment]
I think they just want us to use the w11

**Reply by qingzheng-web** [reply]
haha!You're right.

**Comment by colin4k** [comment]
Change to the following format, then it works:
```
{
"networkingMode": "bridged",
"vmSwitch": "WSL_external"
}
```

**Reply by mrzhang76** [reply]
I think it still does not work after writing to .wslconfig . Could you give me more message about how to use this format? thank you !

**Reply by colin4k** [reply]
> I think it still does not work after writing to .wslconfig . Could you give me more message about how to use this format? thank you !

in .wlsconfig, it should be:
```
{
"networkingMode": "bridged",
"vmSwitch": "WSL",
"dhcp": "false",
"ipv6": "true",
"localhostForwarding": "true"
}
```
then you need to create a file with the path:/usr/lib/systemd/network/wsl_external.network, then input:
```
[Match]
Name=eth0
[Network]
Description=WSLBridge
DHCP=false
Address=192.168.1.222/24
Gateway=192.168.1.1
DNS=192.168.1.1
```
then stop wsl by typing ' wsl --shutdown' , and then enter wsl again, it would be ok.

**Reply by colin4k** [reply]
And my vmswitch is like the following:
```
PS C:\Users\colin> Get-VMSwitch

Name           SwitchType NetAdapterInterfaceDescription
----           ---------- ------------------------------
Default Switch Internal
WSL            External   Intel(R) Ethernet Connection (2) I219-V
```

**Reply by colin4k** [reply]
Oh, I forgot another important step, you should edit /etc/wsl.conf:
```
[boot]
systemd=true
[network]
generateResolvConf = false
```
then,
```
rm -f /etc/resolv.conf
echo "nameserver 192.168.1.1" > /etc/resolv.conf

systemctl list-unit-files --type=service
systemctl enable systemd-networkd
systemctl restart systemd-networkd
systemctl enable systemd-resolved.service
systemctl restart systemd-resolved.service

```

**Reply by jieny** [reply]
> And my vmswitch is like the following:
> 
> ```
> PS C:\Users\colin> Get-VMSwitch
> 
> Name           SwitchType NetAdapterInterfaceDescription
> ----           ---------- ------------------------------
> Default Switch Internal
> WSL            External   Intel(R) Ethernet Connection (2) I219-V
> ```

reboot system then WSL wil be reset to Internal

can use WSL_external? 

The json configuration does not seem to take effect and the specified switch will not be used

**Reply by tylerjin** [reply]
> systemctl restart systemd-resolved.service




> > And my vmswitch is like the following:
> > ```
> > PS C:\Users\colin> Get-VMSwitch
> > 
> > Name           SwitchType NetAdapterInterfaceDescription
> > ----           ---------- ------------------------------
> > Default Switch Internal
> > WSL            External   Intel(R) Ethernet Connection (2) I219-V
> > ```
> 
> reboot system then WSL wil be reset to Internal
> 
> can use WSL_external?
> 
> The json configuration does not seem to take effect and the specified switch will not be used

Run hyper-V as admin then change it to External
![image](https://github.com/microsoft/WSL/assets/14880055/98359dc5-66a0-4f85-ba79-a9adfed3744d)

**Reply by chabad360** [reply]
I've been getting the following error when using the json format:
```
wsl: Invalid key name in C:\Users\***\.wslconfig:1
wsl: Invalid key name in C:\Users\***\.wslconfig:2
wsl: Invalid key name in C:\Users\***\.wslconfig:3
wsl: Invalid key name in C:\Users\***\.wslconfig:4
wsl: Invalid key name in C:\Users\***\.wslconfig:5
wsl: Invalid key name in C:\Users\***\.wslconfig:6
wsl: Invalid key name in C:\Users\***\.wslconfig:7
```

**Reply by Biswa96** [reply]
.wslconfig should be in ini format.

**Reply by chabad360** [reply]
> in .wlsconfig, it should be:
> ```
> {
> "networkingMode": "bridged",
> "vmSwitch": "WSL",
> "dhcp": "false",
> "ipv6": "true",
> "localhostForwarding": "true"
> }
> ```

?

**Reply by The-Monkey-King** [reply]
Is .wslconfig supposed to look like this?

```
[wsl2]
{
"networkingMode": "bridged",
"vmSwitch": "WSL",
"dhcp": "false",
"ipv6": "true",
"localhostForwarding": "true"
}
```

**Reply by electropolis** [reply]
I think .wslconfig should be in ini not json format

**Reply by wuyanze0226** [reply]
it doesn't work. I tried in windows 10

**Comment by elsaco** [comment]
It doesn't work on Windows 10. After creating an external switch named `WSL_external` and adding `{ "networkingMode": "bridged", "vmSwitch": "WSL_external" }` to `.wslconfig` the distro starts w/out errors but networking is ...not working anymore inside the container.

`eth0` is getting the IP from the 172.31.x.x/20 space, still:

```
////////////////////////NETWORKS///////////////////////

Network : C08CB7B8-9B3C-408E-8E30-5E16A3AEB444
    Name             : Default Switch
    Type             : ICS
    Subnet Address   : 172.29.48.0/20
    Gateway          : 172.29.48.1

Network : B95D0C5E-57D4-412B-B571-18A81A16E005
    Name             : WSL
    Type             : ICS
    Subnet Address   : 172.31.128.0/20
    Gateway          : 172.31.128.1

Network : 5599425B-2BD4-4E98-9F2E-FE8E323C733D
    Name             : WSL_external
    Type             : Transparent

Network : E6378EA9-E4C6-4056-B536-CD2F4CAB8CA6
    Name             : vEthernet (WSL_external)
    Type             : ICS
    Subnet Address   : 172.29.224.0/20
    Gateway          : 172.29.224.1

///////////////////////ENDPOINTS///////////////////////

Endpoint         : 62df7b3e-b2db-42af-bddf-48d5e53f3021
    Name             : Ethernet
    IP Address       : 172.31.136.64

//////////////////////NAMESPACES///////////////////////

Namespace        : 910f7d92-ba2d-4c3f-98ae-7c0ac590d2dc

/////////////////////LOADBALANCERS/////////////////////

```

**Reply by qwer1304** [reply]
Anyone managed to get WSL2 work in bridged mode on Windows 10?
If so, could they please provide a detailed complete recepie for how they did it?
Thx

**Reply by safeblood** [reply]
if you got solution please tell me

**Reply by qingzheng-web** [reply]
> Anyone managed to get WSL2 work in bridged mode on Windows 10?有人设法让 WSL2 在 Windows 10 上以桥接模式工作吗？ If so, could they please provide a detailed complete recepie for how they did it?如果是这样，他们能否提供一份详细的完整记录，说明他们是如何做到的？ Thx 感谢

I hope so.

**Comment by Kilill** [comment]
I do not expect WSL2 to ever work correctly on Win10 since MS wants you to migrate to WIN11

**Reply by qingzheng-web** [reply]
It's so sad!

**Comment by shuryanc** [comment]
This works for me. If it doesn't work for you, then figure it out yourselves:

[For Win 10] Edit Hyper-V -> WSL (like what [tylerjin mentioned](https://github.com/microsoft/WSL/discussions/9227#discussioncomment-6036297)) switch to use external
[No need to edit .wslconfig, just leave it blanked]

On WSL machine:
/usr/lib/systemd/network/wsl_external.network

[Match]
Name=eth0
[Network]
Description=WSLBridge
DHCP=false
Address={Your ip for WSL}
Gateway={Gateway of the bridged card}
DNS={DNS used by your bridged card}

/etc/wsl.conf
[boot]
systemd=true
[network]
generateResolvConf = false

/etc/resolv.conf
nameserver {DNS used by your bridged card}
nameserver {DNS used by your bridged card}
nameserver {DNS used by your bridged card}

systemctl enable systemd-networkd
systemctl restart systemd-networkd
systemctl enable systemd-resolved.service
systemctl restart systemd-resolved.service

**Reply by The-Monkey-King** [reply]
Thanks! But I do have a couple of observations and questions.

Observations:
1. To switch the WSL virtual switch to "external," you need to have Hyper-V manager running, then open the Linux instance, refresh Hyper-V manager to have the WSL switch pop up, then terminate the Linux instance. Doing this allows you to see the WSL virtual switch and then allowing you to apply the change, Keeping Linux instance running (for me, Win10) won't let me apply the change.
2. When I applied the change, "Ethernet adapter Ethernet" (my host PC's NIC connection) disappeared. 
3. Rebooting the PC host made my computer act like a normal bridged connection in Windows 10 - that is, the PC lost network connectivity. I believe the WSL virtual switch captured my PC's IP address. And confirming it with 'ipconfig /all' (Windows side), this appears to the case.

Questions:
1. In your wsl_network file (/usr/lib/systemd/network/wsl_external.network), "address" is supposed to be the IP address of what you want your Linux instance to be? Example: Host PC receives a static IP from home router -> 50.10.50.12 :: I want then to have the Linux Instance on host have 50.10.50.22. So would it look like this:
```
[Match]
Name=eth0

[Network]
Description=WSLBridge
DHCP=false
Address=50.10.50.22
Gateway=50.10.50.40
DNS=50.10.50.40
```

3. If the home network router has its DNS server set to 50.10.50.40, is that what I put in the /etc/resolv.conf file?
`nameserver 50.10.50.40 `

4. And do I have to repeat this three times in the /etc/resolv.conf file like you have listed?
```
nameserver 50.10.50.40
nameserver 50.10.50.40
nameserver 50.10.50.40
```

**Comment by Dan010123** [comment]
Hi! 
To switch the WSL to external adapter (win 10 build 19044, WSL 2) 
taking into consideration that the virtual switch for the WSL will be named as WSL by default and my WSL image is ubuntu

1. Created ubuntu.bat for running the WSL as
```
wsl -d ubuntu
powershell -Command "Set-VMSwitch WSL -SwitchType Internal"
```

which sets the WSL virtual switch back to Internal after WSL will be shut down later

2. Made the **shortcut** of this ubuntu.bat and set this shurtcut  to run as admin (create shortcut of the ubuntu.bat and set  -> it's Properties -> Advanced -> Run as Administrator) - simple way to run batch as admin

3. in WSL ubuntu: /etc/wsl.conf made as folowing (my external adapter is named Ethernet)
```
[boot]
systemd=true
command = "/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/powershell.exe -Command 'Set-VMSwitch -name "WSL" -NetAdapterName "Ethernet" -AllowManagementOS $true'; ip addr flush dev eth0; dhclient eth0"
[network]
generateResolvConf = false
```

4. file /etc/resolv.conf (8.8.8.8 or whatever your nameserver is, I used 8.8.8.8) is as following:
` nameserver 8.8.8.8`

now running **shortcut** (not the ubuntu.bat!) will launch WSL, WSL will apply settings from the /etc/wsl.conf = for me gets ip from the router connected to the Ethernet adapter (in bullet 3) and the nameserwer is 8.8.8.8 (in bullet 4). Could be checked with
`hostname -I` 
which gets for me local address inside LAN as 192.168..... 
and  
`ping google.com`
to test that the nameserver works

After WSL shuts down (sudo shutdown now) the script in bullet 1 continues with setting back WSL virtual adapter to Internal (line powershell -Command "Set-VMSwitch WSL -SwitchType Internal")  that prevents the system from loosing net configuration after rebooting the host PC.

that's all for me.

**Comment by Dan010123** [comment]
or without this scripting
1 launch your WSL as you do always, wait for ready state
2 in windows launch powershell in admin mode
3 in powershell type (Ethernet is my external adapter name, WSL - created by the system virtual switch - in image below)
 
![Capture](https://github.com/microsoft/WSL/assets/134538828/77d066da-2b99-4c35-bdea-831861f209b8)

`Set-VMSwitch -name "WSL" -NetAdapterName "Ethernet" -AllowManagementOS $true`
4 in ubuntu 
```
sudo ip addr flush dev eth0
sudo dhclient eth0
```
5. edit /etc/resolv.conf as
 nameserver 8.8.8.8
6. test in ubuntu with hostname -I and ping google.com
7. after WSL shuts down - in powershell enter 
`Set-VMSwitch WSL -SwitchType Internal`

**Reply by athalas** [reply]
important to note: this command must be issued with the VM (WSL2) powered off.

This solution (and the one from you from above) works for me as well, thanks a lot!

**Reply by lozn00** [reply]
I am unstable after using it, and it will automatically recover after restarting the computer. I also need to remove hyper-v  at ethernet attr and re-enter the powshell to execute commands
So, how to write a startup script to enable WSL with just one click after booting up, remove Hyper-V from Ethernet with just one click, and then enable WSL management with just one click

**Comment by 4-FLOSS-Free-Libre-Open-Source-Software** [comment]
Sad news from just released [WSL v2.4.5](https://github.com/microsoft/WSL/releases/tag/2.4.5) changelog:

> Add user warning letting users know bridged networking will be deprecated

Bridged mode works very well as it is now in WSL-Version: 2.4.5.0

**Comment by elsaco** [comment]
The exact message is that `bridged networking mode has been deprecated`:

```
wsl: Bridged networking mode has been deprecated, please see https:://aka.ms/wslnetworking for information about supported networking modes
wsl: Bridged networking requires wsl2.vmSwitch to be set.

```

**Reply by 4-FLOSS-Free-Libre-Open-Source-Software** [reply, marked as answer]
Fortunately the latest pre-release reads

> Bring back bridged networking mode

[2.5.6](https://github.com/microsoft/WSL/releases/tag/2.5.6)

**Reply by osde8info** [reply]
YAY !
